contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,786 @@
+        require(shares > 0, "INVALID_SHARES_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory srcValInfo = _validators[srcValidator];
+        Validator memory dstValInfo = _validators[dstValidator];
+        require(
+            IStakeCredit(srcValInfo.creditContract).getPooledBNBByShares(shares) >= minDelegationBNBChange,
+            "INVALID_REDELEGATION_AMOUNT"
+        );
+
+        if (dstValInfo.jailed) {
+            // only self delegation
+            require(delegator == dstValidator, "ONLY_SELF_DELEGATION");
+        }
+
+        uint256 bnbAmount = IStakeCredit(srcValInfo.creditContract).unbond(delegator, shares);
+        uint256 newShares = IStakeCredit(dstValInfo.creditContract).delegate{ value: bnbAmount }(delegator);
+        emit Redelegated(srcValidator, dstValidator, delegator, shares, newShares, bnbAmount);
+
+        if (delegator == srcValidator) {
+            _checkValidatorSelfDelegation(srcValidator);
+        }
+
+        address[] memory _pools = new address[](2);
+        _pools[0] = srcValInfo.creditContract;
+        _pools[1] = dstValInfo.creditContract;
+        IGovToken(GOV_TOKEN_ADDR).sync(_pools, delegator);
+        IGovToken(GOV_TOKEN_ADDR).delegateVote(delegator, dstValidator);
+    }
+
+    /**
+     * @dev Claim the undelegated BNB from the pool after unbondPeriod
+     */
+    function claim(
+        address operatorAddress,
+        uint256 requestNumber
+    ) external whenNotPaused validatorExist(operatorAddress) {
+        uint256 bnbAmount = IStakeCredit(_validators[operatorAddress].creditContract).claim(msg.sender, requestNumber);
+        emit Claimed(operatorAddress, msg.sender, bnbAmount);
+    }
+
+    function sync(address[] calldata operatorAddresses, address account) external whenNotPaused {
+        uint256 _length = operatorAddresses.length;
+        address[] memory stakeCredits = new address[](_length);
+        address _pool;
+        for (uint256 i = 0; i < _length; ++i) {
+            _pool = _validators[operatorAddresses[i]].creditContract;
+            require(_pool != address(0), "validator not exist");
+            stakeCredits[i] = _pool;
+        }
+
+        IGovToken(GOV_TOKEN_ADDR).sync(stakeCredits, account);
+    }
+
+    /*----------------- system functions -----------------*/
+    function distributeReward(address consensusAddress) external payable onlyValidatorContract {
+        address operatorAddress = _consensusToOperator[consensusAddress];
+        Validator memory valInfo = _validators[operatorAddress];
+        if (valInfo.creditContract == address(0) || valInfo.jailed) {
+            emit RewardDistributeFailed(operatorAddress, "INVALID_VALIDATOR");
+            return;
+        }
+
+        try IStakeCredit(valInfo.creditContract).distributeReward{ value: msg.value }(valInfo.commission.rate) {
+            emit RewardDistributed(operatorAddress, msg.value);
+        } catch Error(string memory reason) {
```

Comment:
Is it possible to revert here? To me, I dont know what error is possible.

Reply:
Removed